reactioncommerce/reaction
# feat: add createRefund GraphQL mutation

Part of #5158 
Impact: **minor**  
Type: **feature**

A new `createRefund` mutation allows you to create a refund for a particular payment on an order. This was previously a meteor only method. The meteor method will be removed with #5158.

## Breaking changes
None

## Testing
1. Place an order
1. Use the following mutation to create a refund on said order
1. Try to refund more than the charge was, and see that only the remaining amout was refunded

```
createRefund(
  amount: {AMOUNT}!,
  orderId: {ORDER-ID}!,
  paymentId: {PAYMENT-ID}!,
  reason: {Reason}) {
    order {
      {FIELDS-TO-RETREIVE}
    }
  }
```

`amount`, `orderId`, and `paymentId` are required. `reason` is optional. It's set up as a `String` to be used with multiple payment providers, however for `Stripe`, which is our only provider at the moment, they only accept 3 values for reason: `duplicate`, `fraudulent`, or `requested_by_customer`, so be sure to use one of those values if testing with Stripe. This limit will be built into the Operator UI if Stripe is the payment method. From their docs:

> String indicating the reason for the refund. If set, possible values are duplicate, fraudulent, and requested_by_customer. If you believe the charge to be fraudulent, specifying fraudulent as the reason will add the associated card and email to your block lists, and will also help us improve our fraud detection algorithms.


I've added tests here to test the permission / requirements of variables, but did not add tests to check a result, as the payment method has it's own tests for this element: https://github.com/reactioncommerce/reaction/blob/feat-kieckhafer-listRefundsGQLQuery/imports/plugins/included/payments-stripe/server/no-meteor/util/stripeListRefunds.test.js#L10
